NEXT-24647 - add MultiInventory page (#829)
shopware/docs: guides/plugins/plugins/api/multi-inventory.md

### ERP System as Single-Source-of-Truth

MultiInventory is intended to be used as an interface between Shopware and your resource management software. This means that Shopware will only calculate availability of Products based on your Warehouse configuration, and only changes stocks of a Warehouse, when an order is created. This is to prevent oversales, while also making your system the single source of truth - making it easier to maintain both systems at the same time.

### Product Availability

Availability of Products is defined in 2 steps

* WarehouseGroups can be assigned to Rules (Rule builder)
  * If the rule is invalid, this Group will not be considered in calculating Product availability.
  * Products / Warehouses still can be available via other groups.
* Products can have a stock per Warehouse
  * All Warehouses inside an active WarehouseGroup are taken into account for calculating the total stock of a specific Product.
  * Warehouses are unique, but can be assigned to multiple Groups (e.g. all Warehouses in the Group "Germany" can also be in the Group "Europe").

> If both conditions are true (e.g. "Customer is in a specific customer group" and "requested stock <= total Product stock of all valid Warehouses"), the requested Product is considered available.

Reviewer: The stock availability is also compatible with other product configurations like maxPurchase, minPurchase, purchaseSteps, etc.

# Multi-Inventory

## Pre-requisites and setup

### Commercial Plugin

> As Multi-Inventory is part of the Commercial plugin, it requires an existing Shopware 6 installation and the activated Commercial plugin on top. The plugin can be installed following these [install instructions](../../../../guides/plugins/plugins/plugin-base-guide#install-your-plugin). In addition, the Multi-Inventory feature requires [Shopware Beyond](https://docs.shopware.com/en/shopware-6-en/features/shopware-beyond).

Author: I would keep it this way "Multi-Inventory is part of the Commercial plugin", since it really is just a specific feature inside the plugin, along with other unrelated features. Applied the rest. 👍🏻

### Admin UI

> While this feature is supposed to be used API first, i.e. by ERP systems, it still comes with an user interface for the administration. If you're looking for user documentation, [please head here](https://docs.shopware.com/en/shopware-6-en/extensions).

Author: Changed "[...] to be used as API first" to "[...]" to be used by API first"

        "versionId": "0fa91ce3e96a4bc2be4bd9ce752c3425",
        "warehouseGroups": [{
            "id": "8cf7736855594501aaf86351e147c61e"
        }],
        "warehouses": [{
            "id": "f5c850109fe64c228377cbd369903b75",
            "productId": "86d38702be7e4ac9a941583933a1c6f5",
            "productVersionId": "0fa91ce3e96a4bc2be4bd9ce752c3425",
            "warehouseId":"4ce2bd36d2824153812fcb6a97f22d22",
            "stock": 0
        }]
    }]
}]
```

### Updating ProductWarehouse stocks

You can update `product_warehouse.stock` in a batch via SyncApi, or patch a specific entity directly via entity repository.

```json

// POST /api/_action/sync

[{
    "action": "upsert",
    "entity": "product_warehouse",
    "payload": [{
        "id": "f5c850109fe64c228377cbd369903b75",
        "stock": 1500
    }, {
        "id": "228377cbd369903b75f5c850109fe64c",
        "stock": 0
    }]
}]

// PATCH /api/product-warehouse/f5c850109fe64c228377cbd369903b75

{
    "id": "f5c850109fe64c228377cbd369903b75",
    "stock": 1500
}
```

## Concept

### ERP System as Single-Source-of-Truth

* Intended to be used via API and custom-built adapter

> * Recommended update frequency?

Author: These bullet points were some ideas, while we were still figuring out what to include (PR was considered a draft). Just updated this section

Multi-Inventory is intended to be used as an interface between Shopware and your resource management software. This means that Shopware will only calculate availability of Products based on your Warehouse configuration, and only changes stocks of a Warehouse, when an order is created. This is to prevent oversales, while also making your system the single source of truth - making it easier to maintain both systems at the same time.

### Product Availability

Availability of Products is defined in 2 steps

* WarehouseGroups can be assigned to Rules (Rule builder)
  * If the rule is invalid, this Group will not be considered in calculating Product availability.

>   * Products / Warehouses still can be available via other groups.

Reviewer 2: ```suggestion
  * Products/Warehouses can still be available via other groups.
```

* Shopware 6 instance
* [Shopware Beyond](https://docs.shopware.com/en/shopware-6-en/features/shopware-beyond) license
* Activated Commercial plugin. Refer to [plugin base guide](../../../../guides/plugins/plugins/plugin-base-guide#install-your-plugin) for installation instructions

### Admin UI

While this feature is supposed to be used by API first, i.e. by ERP systems, it still comes with an user interface for the Administration. Refer to [My extensions](https://docs.shopware.com/en/shopware-6-en/extensions/myextensions) section of shopware docs to explore more about it.

### Admin API

To create, modify or delete Warehouses, WarehouseGroups,etc, related to Multi-Inventory, you can access Admin API endpoints decscribed further.

Meanwhile, refer to the following links regarding the general use of the Admin API:

* [Authentication & Authorization](https://shopware.stoplight.io/docs/admin-api/ZG9jOjEwODA3NjQx-authentication)
* [Request & Response Structure](https://shopware.stoplight.io/docs/admin-api/ZG9jOjEyMzAzNDU1-request-and-response-structure)
* [Endpoint Structure](https://shopware.stoplight.io/docs/admin-api/ZG9jOjEyMzA1ODA5-endpoint-structure)

> ---

Reviewer 2: ```suggestion
```

## Working with the API

The following examples contain payloads for typical use-cases of this feature. Basically all new entities fully support the Admin API via sync service or their generic entity endpoints.

> ### Creating or Updating a WarehouseGroup and assigning it to an existing Warehouse

Reviewer 2: ```suggestion
### Creating or updating a WarehouseGroup and assigning it to an existing Warehouse
```

```json

// POST /api/warehouse-group
// PATCH /api/warehouse-group/8cf7736855594501aaf86351e147c61e

{
    "id": "8cf7736855594501aaf86351e147c61e",
    "name": "Group A",
    "description": "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore.",
    "priority": 25,
    "ruleId": "93248b220a064424a1f6e90010820ba2",
    "warehouses":  [{
        "id": "4ce2bd36d2824153812fcb6a97f22d22"
    }]
}
```

> ### Creating or Updating a Warehouse and assigning it to existing WarehouseGroups

Reviewer 2: ```suggestion
### Creating or updating a Warehouse and assigning it to an existing WarehouseGroups
```

[{
    "action": "upsert",
    "entity": "product",
    "payload": [{
        "id": "86d38702be7e4ac9a941583933a1c6f5",
        "versionId": "0fa91ce3e96a4bc2be4bd9ce752c3425",
        "warehouseGroups": [{
            "id": "8cf7736855594501aaf86351e147c61e"
        }],
        "warehouses": [{
            "id": "f5c850109fe64c228377cbd369903b75",
            "productId": "86d38702be7e4ac9a941583933a1c6f5",
            "productVersionId": "0fa91ce3e96a4bc2be4bd9ce752c3425",
            "warehouseId":"4ce2bd36d2824153812fcb6a97f22d22",
            "stock": 0
        }]
    }]
}]
```

### Updating ProductWarehouse stocks

You can update `product_warehouse.stock` in batch via SyncApi, or patch a specific entity directly via entity repository.

```json

// POST /api/_action/sync

[{
    "action": "upsert",
    "entity": "product_warehouse",
    "payload": [{
        "id": "f5c850109fe64c228377cbd369903b75",
        "stock": 1500
    }, {
        "id": "228377cbd369903b75f5c850109fe64c",
        "stock": 0
    }]
}]

// PATCH /api/product-warehouse/f5c850109fe64c228377cbd369903b75

{
    "id": "f5c850109fe64c228377cbd369903b75",
    "stock": 1500
}
```

## Concept

> Note: Every described behavior only applies to Products, which are assigned to WarehouseGroups - every unrelated Product will use the default Shopware behavior.

Reviewer 2: ```suggestion
{% hint style="info" %}
Every described behavior only applies to Products that are assigned to WarehouseGroups and every unrelated Product will use the default Shopware behavior.
{% endhint %}
```
